### PR TITLE
Add transform support for JSON modules imports

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/2866
| Any Dependency Changes?  |
| License                  | MIT

https://github.com/tc39/proposal-json-modules has been in Stage 3 for three years and we do not support it yet :)

Now that we already have some logic to converts imports to `fetch`/`fs.readFile` for `import source`, we can just reuse it. The first commit extracts that logic to a shared helper; the second commit implements transform for `import ... with { type: "json" }`.

### CODE REVIEW

File: packages/babel-plugin-proposal-json-modules/test/fixtures/browser/import-expression/output.mjs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// TODO: Not supported yet
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-json-modules/test/fixtures/browser/import-expression/output.mjs:1]: I'm not sure how to support this, given that:
- if it's not a JSON import, we need to fallback to an import with two arguments
- that is not known statically, so we need to always emit code for that fallback
- the browser might only support import with a single argument

I'd not block this feature on solving that problem, but rather document it as a known limitation.

**Review by liuxingbaoyu — APPROVED** [review]

File: packages/babel-helper-import-to-platform-api/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+          t.callExpression(t.identifier("fetch"), [imrWithFallback(specifier)]),
+        );
+      break;
+    case p({ web: false, node: true, toCJS: true }):
+      buildFetchSync = specifier =>
+        transformers.nodeFsSync(template.expression.ast`
+            require("fs").readFileSync(require.resolve(${specifier}))
+          `);
+      buildFetchAsync = specifier => template.expression.ast`
+            require("fs").promises.readFile(require.resolve(${specifier}))
```
Comment (JLHwung) [inline packages/babel-helper-import-to-platform-api/src/index.ts:134]: The fs Promises API is available on Node.js 10. If we would like to support older versions, we should manually promisify the callback-style `fs.readFile` call. Ideally we can output the most compatible form based on the input target.

**Review by JLHwung — APPROVED** [review]
The transform looks good to me.

Can you add a test case of integration between this plugin and polyfill providers? For example the web output depends on the fetch API, if users specify `chrome: 41` as target, the `Fetch` API should be polyfilled then.

As for docs, can you also add docs for the `proposal-import-wasm-source` plugin?

**Comment by nicolo-ribaudo** [comment]
> As for docs, can you also add docs for the `proposal-import-wasm-source` plugin?

https://babeljs.io/docs/babel-plugin-proposal-import-wasm-source 😛

**Comment by nicolo-ribaudo** [comment]
> Can you add a test case of integration between this plugin and polyfill providers? For example the web output depends on the fetch API, if users specify chrome: 41 as target, the Fetch API should be polyfilled then.

core-js does not polyfill `fetch`, but I added a test showing that it polyfills `URL` and `Promise`.

**Comment by what1s1ove** [comment]
Hey @nicolo-ribaudo ! Thank you for all your hard work! ❤️

Quick question (didn't find it anywhere). Are there any plans for `with { type:  "css" }`? According to the specification, it should become a [CSSStyleSheet](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet) object.

**Comment by what1s1ove** [comment]
> Hey! We usually only release "official" plugins for features defined in the TC39 standard or proposals to it (+ TypeScript/JSX), and not for features that only work in browsers.
> 
> However, it should be quite easy to copy the plugin for `type: "json"` (https://github.com/babel/babel/blob/main/packages/babel-plugin-proposal-json-modules/src/index.ts) and adapt it to work with `type: "css"`, with the additional simplification that since this would be only for browser code you don't need to also support CommonJS :) I'd encourage you to try, and feel free to ask if you need any help!

Thank you Nicolò ☺️! I will try!

**Pull request merged**